scenarios/two_party_countersigning/src/main.rs

```diff
@@ -0,0 +1,390 @@
+                // As an initiator we just need to call a zome so that `init` will run.
+                client
+                    .call_zome(
+                        app_port,
+                        cell_id,
+                        "countersigning",
+                        "initiator_hello",
+                        (),
+                        None,
+                    )
+                    .await?;
+            } else if assigned_behaviour == "participate" {
+                // As a participant we need to advertise our role by publishing a link to our agent key
+                client
+                    .call_zome(
+                        app_port,
+                        cell_id,
+                        "countersigning",
+                        "participant_hello",
+                        (),
+                        None,
+                    )
+                    .await?;
+            } else {
+                return Err(anyhow::anyhow!(
+                    "Unknown assigned behaviour: {}",
+                    assigned_behaviour
+                ));
+            }
+
+            Ok(())
+        })?;
+
+    log::debug!("Agent setup complete for: {}", ctx.agent_name().to_string());
+
+    Ok(())
+}
+
+fn agent_behaviour_initiate(
+    ctx: &mut AgentContext<TryCPRunnerContext, TryCPAgentContext<ScenarioValues>>,
+) -> HookResult {
+    let client = ctx.get().trycp_client();
+
+    let app_port = ctx.get().app_port();
+    let cell_id = ctx.get().cell_id();
+    let initiate_with_peers = ctx.get_mut().scenario_values.initiate_with_peers.pop();
+    let reporter = ctx.runner_context().reporter();
+
+    let agent_name = ctx.agent_name().to_string();
+    let initiated = ctx.get().scenario_values.session_attempts.clone();
+    let initiated_success = ctx.get().scenario_values.session_successes.clone();
+
+    let new_peers = ctx
+        .runner_context()
+        .executor()
+        .execute_in_place(
+            async move {
+                match initiate_with_peers {
+                    None => {
+                        // No more agents available to call, get a new list.
+                        // This is also the initial condition.
+                        let response = client
+                            .call_zome(
+                                app_port,
+                                cell_id.clone(),
+                                "countersigning",
+                                "list_participants",
+                                (),
+                                None,
+                            )
+                            .await
+                            .context("Failed to list participants")?;
+                        let mut new_peer_list = response.decode::<Vec<AgentPubKey>>().map_err(|e| anyhow::anyhow!("Decoding failure: {:?}", e))?;
+                        new_peer_list.shuffle(&mut thread_rng());
+                        Ok(new_peer_list)
```

Comment:
Just checking my understanding: while we're still setting up waiting for participant links to be published to us, this will effectively be a busy loop running the list_participants zome call as fast as possible?

Reply:
It will indeed, added a sleep to slow it down when no agents are found.